File: JDK11/opentraceability/src/main/java/opentraceability/mappers/epcis/json/epcisjsonmasterdatawriter.java
package opentraceability.mappers.epcis.json;

import Newtonsoft.Json.Linq.*;
import opentraceability.interfaces.*;
import opentraceability.models.common.*;
import opentraceability.models.events.*;
import opentraceability.utility.attributes.*;
import opentraceability.*;
import opentraceability.mappers.*;
import opentraceability.mappers.epcis.*;
import java.util.*;

/** 
 Used for writing master data into an EPCIS JSON-LD file.
*/
public final class EPCISJsonMasterDataWriter
{
	public static void WriteMasterData(JObject jDoc, EPCISBaseDocument doc)
	{
		if (!doc.getMasterData().isEmpty())
		{
			JObject xEPCISHeader = jDoc["epcisHeader"] instanceof JObject ? (JObject)jDoc["epcisHeader"] : null;
			if (xEPCISHeader == null)
			{
				jDoc["epcisHeader"] = new JObject(new JProperty("epcisMasterData", new JObject(new JProperty("vocabularyList", new JArray()))));
			}
			else
			{
				xEPCISHeader["epcisMasterData"] = new JObject(new JProperty("vocabularyList", new JArray()));
			}
			Object tempVar = ((jDoc["epcisHeader"]["epcisMasterData"] == null ? null : jDoc["epcisHeader"]["epcisMasterData"]["vocabularyList"]));
//C# TO JAVA CONVERTER TASK: Throw expressions are not converted by C# to Java Converter:
//ORIGINAL LINE: JArray jVocabList = jDoc["epcisHeader"] == null ? null : tempVar instanceof JArray ? (JArray)tempVar : null ?? throw new Exception("Failed to grab json object epcisHeader.epcisMasterData.vocabularyList");
			JArray jVocabList = jDoc["epcisHeader"] == null ? null : tempVar instanceof JArray ? (JArray)tempVar : (null != null ? null : throw new RuntimeException("Failed to grab json object epcisHeader.epcisMasterData.vocabularyList"));

			for (var mdList : doc.getMasterData().GroupBy(m -> m.EPCISType))
			{
				if (mdList.Key != null)
				{
					WriteMasterDataList(mdList.ToList(), jVocabList, mdList.Key);
				}
				else
				{
					throw new RuntimeException("There are master data vocabulary elements where the Type is NULL.");
				}
			}
		}
	}

	private static void WriteMasterDataList(ArrayList<IVocabularyElement> data, JArray xVocabList, String type)
	{
		if (!data.isEmpty())
		{
			JObject jVocab = new JObject(new JProperty("type", type), new JProperty("vocabularyElementList", new JArray()));
//C# TO JAVA CONVERTER TASK: Throw expressions are not converted by C# to Java Converter:
//ORIGINAL LINE: JArray xVocabEleList = jVocab["vocabularyElementList"] instanceof JArray ? (JArray)jVocab["vocabularyElementList"] : null ?? throw new Exception("Failed to grab the array vocabularyElementList");
			JArray xVocabEleList = jVocab["vocabularyElementList"] instanceof JArray ? (JArray)jVocab["vocabularyElementList"] : (null != null ? null : throw new RuntimeException("Failed to grab the array vocabularyElementList"));

			for (IVocabularyElement md : data)
			{
				JObject xMD = WriteMasterDataObject(md);
				xVocabEleList.Add(xMD);
			}

			xVocabList.Add(jVocab);
		}
	}

	private static JObject WriteMasterDataObject(IVocabularyElement md)
	{
		JObject jVocabElement = new JObject(new JProperty("id", md.getID() != null ? md.getID() : ""), new JProperty("attributes", new JArray()));
//C# TO JAVA CONVERTER TASK: Throw expressions are not converted by C# to Java Converter:
//ORIGINAL LINE: JArray jAttributes = jVocabElement["attributes"] instanceof JArray ? (JArray)jVocabElement["attributes"] : null ?? throw new Exception("Failed to grab attributes array.");
		JArray jAttributes = jVocabElement["attributes"] instanceof JArray ? (JArray)jVocabElement["attributes"] : (null != null ? null : throw new RuntimeException("Failed to grab attributes array."));

		var mappings = OTMappingTypeInformation.GetMasterDataXmlTypeInfo(md.getClass());

		for (var mapping : mappings.getProperties())
		{
			String id = mapping.getName();
			PropertyInfo p = mapping.getProperty();

//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: object? o = p.GetValue(md);
			Object o = p.GetValue(md);
			if (o != null)
			{
				if (Objects.equals(id, ""))
				{
					var subMappings = OTMappingTypeInformation.GetMasterDataXmlTypeInfo(o.getClass());
					for (var subMapping : subMappings.getProperties())
					{
						String subID = subMapping.getName();
						PropertyInfo subProperty = subMapping.getProperty();
//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: object? subObj = subProperty.GetValue(o);
						Object subObj = subProperty.GetValue(o);
						if (subObj != null)
						{
							if (subObj.getClass() == ArrayList<LanguageString>.class)
							{
								ArrayList<LanguageString> l = (ArrayList<LanguageString>)subObj;
//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: string? str = l.FirstOrDefault() == null ? null : l.FirstOrDefault().Value;
								String str = l.FirstOrDefault() == null ? null : l.FirstOrDefault().Value;
								if (str != null)
								{
									JObject jAttribute = new JObject(new JProperty("id", subID), new JProperty("attribute", str));
									jAttributes.Add(jAttribute);
								}
							}
							else
							{
//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: string? str = subObj.ToString();
								String str = subObj.toString();
								if (str != null)
								{
									JObject jAttribute = new JObject(new JProperty("id", subID), new JProperty("attribute", str));
									jAttributes.Add(jAttribute);
								}
							}
						}
					}
				}
				else if (p.<OpenTraceabilityObjectAttribute>GetCustomAttribute() != null)
				{
					JObject jAttribute = new JObject(new JProperty("id", id), new JProperty("attribute", WriteObject(p.PropertyType, o)));
					jAttributes.Add(jAttribute);
				}
				else if (p.<OpenTraceabilityArrayAttribute>GetCustomAttribute() != null)
				{
					List l = (List)o;
					for (var i : l)
					{
//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: string? str = i.ToString();
						String str = i.toString();
						if (str != null)
						{
							JObject jAttribute = new JObject(new JProperty("id", id), new JProperty("attribute", str));
							jAttributes.Add(jAttribute);
						}
					}
				}
				else if (o.getClass() == ArrayList<LanguageString>.class)
				{
					ArrayList<LanguageString> l = (ArrayList<LanguageString>)o;
//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: string? str = l.FirstOrDefault() == null ? null : l.FirstOrDefault().Value;
					String str = l.FirstOrDefault() == null ? null : l.FirstOrDefault().Value;
					if (str != null)
					{
						JObject jAttribute = new JObject(new JProperty("id", id), new JProperty("attribute", str));
						jAttributes.Add(jAttribute);
					}
				}
				else
				{
//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: string? str = o.ToString();
					String str = o.toString();
					if (str != null)
					{
						JObject jAttribute = new JObject(new JProperty("id", id), new JProperty("attribute", str));
						jAttributes.Add(jAttribute);
					}
				}
			}
		}

		for (IMasterDataKDE kde : md.getKDEs())
		{
			JToken jKDE = kde.GetGS1WebVocabJson();
			if (jKDE != null)
			{
				JObject jAttribute = new JObject(new JProperty("id", kde.getName()), new JProperty("attribute", jKDE));
				jAttributes.Add(jAttribute);
			}
		}

		return jVocabElement;
	}

	private static JObject WriteObject(java.lang.Class t, Object o)
	{
		JObject j = new JObject();
		for (var property : t.GetProperties())
		{
//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: object? value = property.GetValue(o);
			Object value = property.GetValue(o);
			if (value != null)
			{
//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: OpenTraceabilityAttribute? xmlAtt = property.GetCustomAttribute<OpenTraceabilityAttribute>();
				OpenTraceabilityAttribute xmlAtt = property.<OpenTraceabilityAttribute>GetCustomAttribute();
				if (xmlAtt != null)
				{
					if (property.<OpenTraceabilityObjectAttribute>GetCustomAttribute() != null)
					{
						j[xmlAtt.getName()] = WriteObject(property.PropertyType, value);
					}
					else
					{
//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: string? str = value.ToString();
						String str = value.toString();
						if (str != null)
						{
							j[xmlAtt.getName()] = str;
						}
					}
				}
			}
		}
		return j;
	}
}

File: CSharp/OpenTraceability/Mappers/EPCIS/JSON/EPCISJsonMasterDataWriter.cs
using Newtonsoft.Json.Linq;
using OpenTraceability.Interfaces;
using OpenTraceability.Models.Common;
using OpenTraceability.Models.Events;
using OpenTraceability.Utility.Attributes;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;

namespace OpenTraceability.Mappers.EPCIS.JSON
{
    /// <summary>
    /// Used for writing master data into an EPCIS JSON-LD file.
    /// </summary>
    public static class EPCISJsonMasterDataWriter
    {
        public static void WriteMasterData(JObject jDoc, EPCISBaseDocument doc)
        {
            if (doc.MasterData.Count > 0)
            {
                JObject xEPCISHeader = jDoc["epcisHeader"] as JObject;
                if (xEPCISHeader == null)
                {
                    jDoc["epcisHeader"] = new JObject(new JProperty("epcisMasterData", new JObject(new JProperty("vocabularyList", new JArray()))));
                }
                else
                {
                    xEPCISHeader["epcisMasterData"] = new JObject(new JProperty("vocabularyList", new JArray()));
                }
                JArray jVocabList = jDoc["epcisHeader"]?["epcisMasterData"]?["vocabularyList"] as JArray ?? throw new Exception("Failed to grab json object epcisHeader.epcisMasterData.vocabularyList");

                foreach (var mdList in doc.MasterData.GroupBy(m => m.EPCISType))
                {
                    if (mdList.Key != null)
                    {
                        WriteMasterDataList(mdList.ToList(), jVocabList, mdList.Key);
                    }
                    else
                    {
                        throw new Exception("There are master data vocabulary elements where the Type is NULL.");
                    }
                }
            }
        }

        private static void WriteMasterDataList(List<IVocabularyElement> data, JArray xVocabList, string type)
        {
            if (data.Count > 0)
            {
                JObject jVocab = new JObject(new JProperty("type", type), new JProperty("vocabularyElementList", new JArray()));
                JArray xVocabEleList = jVocab["vocabularyElementList"] as JArray ?? throw new Exception("Failed to grab the array vocabularyElementList");

                foreach (IVocabularyElement md in data)
                {
                    JObject xMD = WriteMasterDataObject(md);
                    xVocabEleList.Add(xMD);
                }

                xVocabList.Add(jVocab);
            }
        }

        private static JObject WriteMasterDataObject(IVocabularyElement md)
        {
            JObject jVocabElement = new JObject(new JProperty("id", md.ID ?? string.Empty), new JProperty("attributes", new JArray()));
            JArray jAttributes = jVocabElement["attributes"] as JArray ?? throw new Exception("Failed to grab attributes array.");

            var mappings = OTMappingTypeInformation.GetMasterDataXmlTypeInfo(md.GetType());

            foreach (var mapping in mappings.Properties)
            {
                string id = mapping.Name;
                PropertyInfo p = mapping.Property;

                object o = p.GetValue(md);
                if (o != null)
                {
                    if (id == string.Empty)
                    {
                        var subMappings = OTMappingTypeInformation.GetMasterDataXmlTypeInfo(o.GetType());
                        foreach (var subMapping in subMappings.Properties)
                        {
                            string subID = subMapping.Name;
                            PropertyInfo subProperty = subMapping.Property;
                            object subObj = subProperty.GetValue(o);
                            if (subObj != null)
                            {
                                if (subObj.GetType() == typeof(List<LanguageString>))
                                {
                                    List<LanguageString> l = (List<LanguageString>)subObj;
                                    string str = l.FirstOrDefault()?.Value;
                                    if (str != null)
                                    {
                                        JObject jAttribute = new JObject(new JProperty("id", subID), new JProperty("attribute", str));
                                        jAttributes.Add(jAttribute);
                                    }
                                }
                                else
                                {
                                    string str = subObj.ToString();
                                    if (str != null)
                                    {
                                        JObject jAttribute = new JObject(new JProperty("id", subID), new JProperty("attribute", str));
                                        jAttributes.Add(jAttribute);
                                    }
                                }
                            }
                        }
                    }
                    else if (p.GetCustomAttribute<OpenTraceabilityObjectAttribute>() != null)
                    {
                        JObject jAttribute = new JObject(new JProperty("id", id), new JProperty("attribute", WriteObject(p.PropertyType, o)));
                        jAttributes.Add(jAttribute);
                    }
                    else if (p.GetCustomAttribute<OpenTraceabilityArrayAttribute>() != null)
                    {
                        IList l = (IList)o;
                        foreach (var i in l)
                        {
                            string str = i.ToString();
                            if (str != null)
                            {
                                JObject jAttribute = new JObject(new JProperty("id", id), new JProperty("attribute", str));
                                jAttributes.Add(jAttribute);
                            }
                        }
                    }
                    else if (o.GetType() == typeof(List<LanguageString>))
                    {
                        List<LanguageString> l = (List<LanguageString>)o;
                        string str = l.FirstOrDefault()?.Value;
                        if (str != null)
                        {
                            JObject jAttribute = new JObject(new JProperty("id", id), new JProperty("attribute", str));
                            jAttributes.Add(jAttribute);
                        }
                    }
                    else
                    {
                        string str = o.ToString();
                        if (str != null)
                        {
                            JObject jAttribute = new JObject(new JProperty("id", id), new JProperty("attribute", str));
                            jAttributes.Add(jAttribute);
                        }
                    }
                }
            }

            foreach (IMasterDataKDE kde in md.KDEs)
            {
                JToken jKDE = kde.GetGS1WebVocabJson();
                if (jKDE != null)
                {
                    JObject jAttribute = new JObject(new JProperty("id", kde.Name), new JProperty("attribute", jKDE));
                    jAttributes.Add(jAttribute);
                }
            }

            return jVocabElement;
        }

        private static JObject WriteObject(Type t, object o)
        {
            JObject j = new JObject();
            foreach (var property in t.GetProperties())
            {
                object value = property.GetValue(o);
                if (value != null)
                {
                    OpenTraceabilityAttribute xmlAtt = property.GetCustomAttribute<OpenTraceabilityAttribute>();
                    if (xmlAtt != null)
                    {
                        if (property.GetCustomAttribute<OpenTraceabilityObjectAttribute>() != null)
                        {
                            j[xmlAtt.Name] = WriteObject(property.PropertyType, value);
                        }
                        else
                        {
                            string str = value.ToString();
                            if (str != null)
                            {
                                j[xmlAtt.Name] = str;
                            }
                        }
                    }
                }
            }
            return j;
        }
    }
}
